app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java

```diff
@@ -0,0 +1,146 @@
+package com.nutomic.syncthingandroid.model;
+
+import android.util.Log;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * This class caches remote folder and device synchronization
+ * completion indicators defined in {@link CompletionInfo#CompletionInfo}
+ * according to syncthing's REST "/completion" JSON result schema.
+ * Completion model of syncthing's web UI is completion[deviceId][folderId]
+ */
+public class Completion {
+
+    private static final String TAG = "Completion";
+
+    HashMap<String, HashMap<String, CompletionInfo>> deviceFolderMap =
+        new HashMap<String, HashMap<String, CompletionInfo>>();
+
+    /**
+     * Adds a device to the cache model if it does not exist.
+     */
+    private void addDevice(String deviceId) {
+        if (!deviceFolderMap.containsKey(deviceId)) {
+            deviceFolderMap.put(deviceId, new HashMap<String, CompletionInfo>());
+        }
+    }
+
+    /**
+     * Removes a device from the cache model.
+     */
+    public void removeDevice(String deviceId) {
+        if (deviceFolderMap.containsKey(deviceId)) {
+            deviceFolderMap.remove(deviceId);
+        }
+    }
+
+    /**
+     * Adds a folder to the cache model if it does not exist.
+     */
+    private void addFolder(String deviceId, String folderId) {
+        if (!deviceFolderMap.containsKey(deviceId)) {
+            addDevice(deviceId);
+        }
+
+        if (!deviceFolderMap.get(deviceId).containsKey(folderId)) {
+            deviceFolderMap.get(deviceId).put(folderId, new CompletionInfo());
+        }
+    }
+
+    /**
+     * Removes a folder from the cache model.
+     */
+    public void removeFolder(String folderId) {
+        for (String deviceId : deviceFolderMap.keySet()) {

```

Comment:
I don't think you need half of these methods, of you grt or default and always update from config.

Reply:
I'll put this on my todo list. For now, I've combined the two function Completion.updateDevices and Completion.updateFolders into one, made the code less redundant checking for existing keys on the hash maps. Added detection of newly created devices on configReload event. The detection of newly created folders (with applying them to the mCompletion model) would require two also implement the two nested hashmap approach in all code parts referring to "Nutomics original" Folder.java and Device.java model classes. I'd suggest doing this in a later PR.

Reply:
So, for now updateFromConfig is complete adopting the existing Folder.java model containing the devices in a list under every folder. It checks for added folders since commit https://github.com/syncthing/syncthing-android/pull/1079/commits/f652b265c39fd7eaa5d3fbdbb3d36d82a1db76c5